File: 11.ZAlgo.py
# The Z Algorithm is a string-matching algorithm that finds all 
# occurrences of a pattern inside a text in linear time O(n).
# It works by constructing a Z-array, where each element represents the longest substring starting 
# from that position that matches the prefix of the string.

# Imagine you have a pattern (the word you want to search for) and 
# a text (where you want to find the pattern).
# Instead of comparing the pattern with the text character by character 
# (which is slow), Z Algorithm concatenates the pattern, a special delimiter (like $), and the text.

# It then builds the Z-array, which tells us at each index how many characters match with the prefix.
# Whenever a value in the Z-array is equal to the pattern length, we know that the pattern appears 
# in the text at that position.

def calculateZ(s):
    n = len(s)
    Z = [0] * n
    l, r = 0, 0

    for i in range(1, n):
        if i <= r:
            Z[i] = min(r - i + 1, Z[i - l])

        while i + Z[i] < n and s[Z[i]] == s[i + Z[i]]:
            Z[i] += 1

        if i + Z[i] - 1 > r:
            l, r = i, i + Z[i] - 1

    return Z

def searchPattern(text, pattern):
    concat = pattern + "$" + text
    Z = calculateZ(concat)
    result = []

    for i in range(len(pattern) + 1, len(concat)):
        if Z[i] == len(pattern):
            result.append(i - len(pattern) - 1)

    return result

text = "abxabcabcaby"
pattern  = "abcaby"
positions = searchPattern(text, pattern)

print("Pattern found at indices:", positions)

File: 11.ZAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class Main {

    // Function to build Z-array
    static int[] calculateZ(String s) {
        int n = s.length();                   // Length of the string
        int[] Z = new int[n];                 // Initialize Z-array with zeros
        int l = 0, r = 0;                     // Left and right boundaries of Z-box

        for (int i = 1; i < n; i++) {         // Start from index 1, prefix always at index 0
            if (i <= r) {                     // If i lies inside the Z-box
                Z[i] = Math.min(r - i + 1, Z[i - l]);
            }
            // Try expanding substring starting from i
            while (i + Z[i] < n && s.charAt(Z[i]) == s.charAt(i + Z[i])) {
                Z[i]++;
            }
            // Update Z-box if we expanded past r
            if (i + Z[i] - 1 > r) {
                l = i;
                r = i + Z[i] - 1;
            }
        }
        return Z;                             // Return constructed Z-array
    }

    // Function to search pattern occurrences in text
    static List<Integer> searchPattern(String text, String pattern) {
        String concat = pattern + "$" + text; // Concatenate pattern + special char + text
        int[] Z = calculateZ(concat);         // Generate Z-array for concatenated string
        List<Integer> result = new ArrayList<>(); // To store indices of matches

        for (int i = pattern.length() + 1; i < concat.length(); i++) {
            if (Z[i] == pattern.length()) {   // If Z-value equals length of pattern → match
                result.add(i - pattern.length() - 1); // Store index in text
            }
        }
        return result;                        // Return all positions of matches
    }

    public static void main(String[] args) throws IOException {
        String text = "abxabcabcaby";         // Example text
        String pattern = "abcaby";            // Example pattern

        List<Integer> positions = searchPattern(text, pattern);
        System.out.println("Pattern found at indices: " + positions);
    }
}
